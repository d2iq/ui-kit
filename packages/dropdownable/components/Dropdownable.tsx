import React from "react";
import styled, { css } from "react-emotion";

import Overlay from "../../shared/components/Overlay";
import DropdownContents from "./DropdownContents";
import resizeEventManager from "../../utilities/resizeEventManager";
import { zIndexDropdownable } from "../../design-tokens/build/js/designTokens";

export enum Direction {
  BottomLeft = "bottom-left",
  BottomRight = "bottom-right",
  BottomCenter = "bottom-center",
  TopLeft = "top-left",
  TopRight = "top-right",
  TopCenter = "top-center"
}

const DEFAULT_DIRECTION_PREFERENCES = [
  Direction.BottomLeft,
  Direction.BottomRight,
  Direction.TopRight,
  Direction.TopLeft
];

const Container = styled("div")`
  position: relative;
`;

export interface DropdownableProps {
  open: boolean;
  dropdown: React.ReactElement<any>;
  preferredDirections?: Direction[];
  onClose?: () => void;
  overlayRoot?: HTMLElement;
}

interface PositionCoord {
  top: number;
  left: number;
}

interface State {
  direction: Direction;
  position: PositionCoord;
}

interface Dimension {
  width: number;
  height: number;
}

const METHODS_TO_BIND = [
  "getDropdown",
  "setPosition",
  "setPositionFromCurrentProps",
  "calculateBestDirection",
  "positionForDirection",
  "childBounds",
  "windowDimensions",
  "dropdownDimensions"
];

class Dropdownable extends React.Component<DropdownableProps, State> {
  public static defaultProps: Partial<DropdownableProps> = {
    preferredDirections: DEFAULT_DIRECTION_PREFERENCES
  };

  private child = React.createRef<HTMLDivElement>();
  private dropdown = React.createRef<HTMLDivElement>();

  constructor(props) {
    super(props);

    this.state = {
      direction: Direction.BottomLeft,
      position: {
        top: 0,
        left: 0
      }
    };

    METHODS_TO_BIND.forEach(method => {
      this[method] = this[method].bind(this);
    });
  }

  componentDidMount() {
    resizeEventManager.add(this.setPositionFromCurrentProps);
    this.setPositionFromCurrentProps();
  }

  componentWillUnmount() {
    resizeEventManager.remove(this.setPositionFromCurrentProps);
  }

  UNSAFE_componentWillReceiveProps(nextProps: DropdownableProps) {
    if (nextProps.open) {
      this.setPosition(nextProps);
    }
  }

  render() {
    return (
      <Container>
        <div ref={this.child}>{this.props.children}</div>
        {this.getOverlay()}
      </Container>
    );
  }

  getOverlay() {
    const { position } = this.state;
    const { open, overlayRoot } = this.props;

    const overlayStyle = css({
      top: `${position.top}px`,
      left: `${position.left}px`,
      position: "absolute",
      transform: `scale(${open ? 1 : 0})`,
      opacity: open ? 1 : 0,
      "z-index": zIndexDropdownable
    });
    return (
      <Overlay overlayRoot={overlayRoot} className={overlayStyle}>
        {this.getDropdown()}
      </Overlay>
    );
  }

  getDropdown() {
    const { open, dropdown, onClose } = this.props;

    return (
      <div ref={this.dropdown}>
        <DropdownContents open={open} onClose={onClose}>
          {React.cloneElement(dropdown, {
            direction: this.state.direction
          })}
        </DropdownContents>
      </div>
    );
  }

  setPositionFromCurrentProps() {
    if (!this.props.open) {
      return;
    }
    this.setPosition(this.props);
  }

  setPosition(props: DropdownableProps) {
    const dropdownDimensions: Dimension = this.dropdownDimensions();
    const windowDimensions: Dimension = this.windowDimensions();
    const childBounds: ClientRect = this.childBounds();
    const direction: Direction = this.calculateBestDirection(
      props.preferredDirections,
      childBounds,
      dropdownDimensions,
      windowDimensions
    );

    this.setState({
      direction,
      position: this.positionForDirection(
        direction,
        childBounds,
        dropdownDimensions
      )
    });
  }

  calculateBestDirection(
    preferredDirections,
    childBounds,
    dropdownDimensions,
    windowDimensions
  ): Direction {
    // Determine if there is enough space in each direction to fit dropdown
    const triggerCenter = childBounds.width / 2 + childBounds.left;
    const possibleVertDirections = {
      top: dropdownDimensions.height <= childBounds.top,
      bottom:
        childBounds.bottom + dropdownDimensions.height <=
        windowDimensions.height
    };
    const possibleHorizDirections = {
      left:
        dropdownDimensions.width + childBounds.left <= windowDimensions.width,
      right: childBounds.right - dropdownDimensions.width >= 0,
      center:
        triggerCenter - dropdownDimensions.width / 2 > 0 &&
        triggerCenter + dropdownDimensions.width / 2 < windowDimensions.width
    };

    // Pick the first available preference
    const preferredDirection = preferredDirections.find(direction => {
      const topOrBottom = direction.split("-", 2)[0];
      const leftOrRight = direction.split("-", 2)[1];
      return (
        possibleVertDirections[topOrBottom] &&
        possibleHorizDirections[leftOrRight]
      );
    });

    // If nothing fits, fall back to the first preference
    return preferredDirection || preferredDirections[0];
  }

  positionForDirection(
    direction: Direction,
    childBounds,
    dropdownDimensions
  ): PositionCoord {
    const isTop: boolean =
      direction === Direction.TopLeft ||
      direction === Direction.TopRight ||
      direction === Direction.TopCenter;
    const isLeft: boolean =
      direction === Direction.TopLeft || direction === Direction.BottomLeft;
    const isCenter: boolean =
      direction === Direction.TopCenter || direction === Direction.BottomCenter;

    const getHorizAlignment = () => {
      if (isCenter) {
        return (
          childBounds.x +
          childBounds.width / 2 -
          dropdownDimensions.width / 2 +
          window.scrollX
        );
      } else if (isLeft) {
        return childBounds.left + window.scrollX;
      } else {
        return (
          childBounds.left -
          dropdownDimensions.width +
          childBounds.width +
          window.scrollX
        );
      }
    };

    return {
      top: isTop
        ? childBounds.top - dropdownDimensions.height + window.scrollY
        : childBounds.top + childBounds.height + window.scrollY,
      left: getHorizAlignment()
    };
  }

  childBounds(): ClientRect {
    const child = this.child.current;

    if (child) {
      return child.getBoundingClientRect();
    }

    // child.current should not be null due to React lifecycle but
    // is not guarateed by the type.
    return {
      width: 0,
      height: 0,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0
    };
  }

  windowDimensions(): Dimension {
    const body = document.body;
    return { width: body.clientWidth, height: body.clientHeight };
  }

  dropdownDimensions(): Dimension {
    const dropdown = this.dropdown.current!;

    if (dropdown) {
      return {
        width: dropdown.clientWidth,
        height: dropdown.clientHeight
      };
    }

    // dropdown.current should not be null due to React lifecycle but
    // is not guarateed by the type.
    return {
      width: 0,
      height: 0
    };
  }
}

export default Dropdownable;
