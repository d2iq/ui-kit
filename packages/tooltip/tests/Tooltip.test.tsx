import React from "react";
import { mount } from "enzyme";
import serializer from "@emotion/jest";
import toJson from "enzyme-to-json";
import { Tooltip } from "../";
import Dropdownable, {
  Direction
} from "../../dropdownable/components/Dropdownable";

expect.addSnapshotSerializer(serializer);

describe("Tooltip", () => {
  it("renders", () => {
    const component = mount(
      <Tooltip id="renders" trigger="trigger">
        content
      </Tooltip>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("renders open tooltip", () => {
    const component = mount(
      <Tooltip id="opened" trigger="trigger" isOpen={true}>
        content
      </Tooltip>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("renders with preferred directions", () => {
    const component = mount(
      <Tooltip
        id="customDir"
        trigger="trigger"
        isOpen={true}
        preferredDirections={[Direction.BottomLeft]}
      >
        content
      </Tooltip>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("passes isOpen={true} to Dropdownable on mouseOver", () => {
    const component = mount(
      <Tooltip id="hoverOpen" trigger="trigger">
        content
      </Tooltip>
    );

    expect(component.find(Dropdownable).prop("isOpen")).toBe(false);
    component.simulate("mouseOver");
    expect(component.find(Dropdownable).prop("isOpen")).toBe(true);
  });

  it("passes isOpen={false} to Dropdownable on mouseLeave", () => {
    const component = mount(
      <Tooltip id="hoverOpen" trigger="trigger" isOpen={true}>
        content
      </Tooltip>
    );
    expect(component.find(Dropdownable).prop("isOpen")).toBe(true);
    component.simulate("mouseLeave");
    expect(component.find(Dropdownable).prop("isOpen")).toBe(false);
  });

  it("does not pass a different value to Dropdownable's 'isOpen' prop on mouseEnter if suppress is true", () => {
    const component = mount(
      <Tooltip id="hoverOpen" trigger="trigger" suppress={true}>
        content
      </Tooltip>
    );
    expect(component.find(Dropdownable).prop("isOpen")).toBeFalsy();
    component.simulate("mouseEnter");
    expect(component.find(Dropdownable).prop("isOpen")).toBeFalsy();
  });

  it("calls onClose prop when closed", () => {
    const handleClose = jest.fn();

    const component = mount(
      <Tooltip id="hoverOpen" trigger="trigger" onClose={handleClose}>
        content
      </Tooltip>
    );

    component.simulate("mouseEnter");
    component.simulate("mouseLeave");
    expect(handleClose).toHaveBeenCalledTimes(1);
  });
});
