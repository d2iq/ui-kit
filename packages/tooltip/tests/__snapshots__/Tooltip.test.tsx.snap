// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
<Tooltip
  id="renders"
  trigger="trigger"
>
  <span
    data-cy="tooltip"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <Dropdownable
      dropdown={
        <TooltipContent
          id="renders"
          isOpen={false}
        >
          content
        </TooltipContent>
      }
      isOpen={false}
      preferredDirections={
        Array [
          "top",
          "top-start",
          "top-end",
          "bottom",
          "bottom-start",
          "bottom-end",
        ]
      }
    >
      <div>
        trigger
      </div>
    </Dropdownable>
  </span>
</Tooltip>
`;

exports[`Tooltip renders open tooltip 1`] = `
.emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<Tooltip
  id="opened"
  isOpen={true}
  trigger="trigger"
>
  <span
    data-cy="tooltip"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <Dropdownable
      dropdown={
        <TooltipContent
          id="opened"
          isOpen={true}
        >
          content
        </TooltipContent>
      }
      isOpen={true}
      preferredDirections={
        Array [
          "top",
          "top-start",
          "top-end",
          "bottom",
          "bottom-start",
          "bottom-end",
        ]
      }
    >
      <div>
        trigger
      </div>
      <clickOutside(DropdownContents)
        isOpen={true}
      >
        <DropdownContents
          isOpen={true}
        >
          <ForwardRef
            style={
              Object {
                "left": "0",
                "position": "absolute",
                "top": "0",
                "zIndex": 2,
              }
            }
          >
            <Overlay
              innerRef={[Function]}
              overlayRoot={
                .emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<body>
                  <div>
                    <div
                      style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-hidden="false"
                        class="inverseColorMode emotion-0"
                        data-cy="tooltipContent"
                        id="opened"
                        role="tooltip"
                      >
                        content
                      </div>
                      <div
                        class="emotion-1"
                      />
                    </div>
                  </div>
                </body>
              }
              style={
                Object {
                  "left": "0",
                  "position": "absolute",
                  "top": "0",
                  "zIndex": 2,
                }
              }
            >
              <Portal
                containerInfo={
                  .emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<div>
                    <div
                      style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-hidden="false"
                        class="inverseColorMode emotion-0"
                        data-cy="tooltipContent"
                        id="opened"
                        role="tooltip"
                      >
                        content
                      </div>
                      <div
                        class="emotion-1"
                      />
                    </div>
                  </div>
                }
              >
                <div
                  style={
                    Object {
                      "left": "0",
                      "position": "absolute",
                      "top": "0",
                      "zIndex": 2,
                    }
                  }
                >
                  <TooltipContent
                    id="opened"
                    isOpen={true}
                  >
                    <div
                      aria-hidden={false}
                      className="inverseColorMode emotion-0"
                      data-cy="tooltipContent"
                      id="opened"
                      role="tooltip"
                      style={
                        Object {
                          "maxWidth": undefined,
                          "minWidth": undefined,
                        }
                      }
                    >
                      content
                    </div>
                    <div
                      className="emotion-1"
                    />
                  </TooltipContent>
                </div>
              </Portal>
            </Overlay>
          </ForwardRef>
        </DropdownContents>
      </clickOutside(DropdownContents)>
    </Dropdownable>
  </span>
</Tooltip>
`;

exports[`Tooltip renders with preferred directions 1`] = `
.emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<Tooltip
  id="customDir"
  isOpen={true}
  preferredDirections={
    Array [
      "bottom-start",
    ]
  }
  trigger="trigger"
>
  <span
    data-cy="tooltip"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <Dropdownable
      dropdown={
        <TooltipContent
          id="customDir"
          isOpen={true}
        >
          content
        </TooltipContent>
      }
      isOpen={true}
      preferredDirections={
        Array [
          "bottom-start",
        ]
      }
    >
      <div>
        trigger
      </div>
      <clickOutside(DropdownContents)
        isOpen={true}
      >
        <DropdownContents
          isOpen={true}
        >
          <ForwardRef
            style={
              Object {
                "left": "0",
                "position": "absolute",
                "top": "0",
                "zIndex": 2,
              }
            }
          >
            <Overlay
              innerRef={[Function]}
              overlayRoot={
                .emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<body>
                  <div>
                    <div
                      data-popper-escaped="true"
                      data-popper-placement="top"
                      data-popper-reference-hidden="true"
                      style="z-index: 2; position: absolute; left: 0px; top: 0px; visibility: hidden; pointer-events: none; bottom: 0px; transform: translate(0px, 0px);"
                    >
                      <div
                        aria-hidden="false"
                        class="inverseColorMode emotion-0"
                        data-cy="tooltipContent"
                        id="opened"
                        role="tooltip"
                      >
                        content
                      </div>
                      <div
                        class="emotion-1"
                      />
                    </div>
                  </div>
                  <div>
                    <div
                      style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-hidden="false"
                        class="inverseColorMode emotion-0"
                        data-cy="tooltipContent"
                        id="customDir"
                        role="tooltip"
                      >
                        content
                      </div>
                      <div
                        class="emotion-1"
                      />
                    </div>
                  </div>
                </body>
              }
              style={
                Object {
                  "left": "0",
                  "position": "absolute",
                  "top": "0",
                  "zIndex": 2,
                }
              }
            >
              <Portal
                containerInfo={
                  .emotion-0 {
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  border-radius: 2px;
  box-sizing: border-box;
  font-size: 12px;
  -webkit-transform: translateY(
      -9px
    );
  -moz-transform: translateY(
      -9px
    );
  -ms-transform: translateY(
      -9px
    );
  transform: translateY(
      -9px
    );
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-1 {
  position: absolute;
  bottom: 4px;
  left: calc(50% - 5px);
  background-color: var(--themeBgPrimaryInverted, #1B2029);
  box-sizing: border-box;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  height: 10px;
  width: 10px;
}

<div>
                    <div
                      style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-hidden="false"
                        class="inverseColorMode emotion-0"
                        data-cy="tooltipContent"
                        id="customDir"
                        role="tooltip"
                      >
                        content
                      </div>
                      <div
                        class="emotion-1"
                      />
                    </div>
                  </div>
                }
              >
                <div
                  style={
                    Object {
                      "left": "0",
                      "position": "absolute",
                      "top": "0",
                      "zIndex": 2,
                    }
                  }
                >
                  <TooltipContent
                    id="customDir"
                    isOpen={true}
                  >
                    <div
                      aria-hidden={false}
                      className="inverseColorMode emotion-0"
                      data-cy="tooltipContent"
                      id="customDir"
                      role="tooltip"
                      style={
                        Object {
                          "maxWidth": undefined,
                          "minWidth": undefined,
                        }
                      }
                    >
                      content
                    </div>
                    <div
                      className="emotion-1"
                    />
                  </TooltipContent>
                </div>
              </Portal>
            </Overlay>
          </ForwardRef>
        </DropdownContents>
      </clickOutside(DropdownContents)>
    </Dropdownable>
  </span>
</Tooltip>
`;
